components/brave_wallet/browser/internal/polkadot_extrinsic.rs

```diff
@@ -0,0 +1,274 @@
+    AlreadyUnwrapped,
+    /// The supplied chain name did not match our hard-coded whitelist.
+    ChainNameNotFound,
+    /// Invalid SCALE value found.
+    InvalidScale,
+    /// Invalid metadata such as the wrong pallet index or call index.
+    InvalidMetadata,
+    /// Invalid length.
+    InvalidLength,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::ChainNameNotFound => {
+                write!(f, "The supplied chain spec name did not match the whitelist.")
+            }
+            Error::AlreadyUnwrapped => write!(f, "Already unwrapped."),
+            Error::InvalidScale => write!(f, "Invalid SCALE-encoded bytes were found."),
+            Error::InvalidMetadata => write!(f, "Invalid chain metadata was encountered."),
+            Error::InvalidLength => write!(f, "Invalid length."),
+        }
+    }
+}
+
+struct CxxPolkadotChainMetadata {
+    balances_pallet_index: u8,
+    transfer_allow_death_call_index: u8,
+}
+
+struct CxxPolkadotChainMetadataResult(Result<CxxPolkadotChainMetadata, Error>);
+impl_result!(CxxPolkadotChainMetadata, CxxPolkadotChainMetadataResult);
+
+struct CxxPolkadotDecodeUnsignedTransferResult(Result<CxxPolkadotDecodeUnsignedTransfer, Error>);
+impl_result!(CxxPolkadotDecodeUnsignedTransfer, CxxPolkadotDecodeUnsignedTransferResult);
+
+fn make_chain_metadata(chain_name: &str) -> Box<CxxPolkadotChainMetadataResult> {
+    let metadata = match chain_name {
+        "Westend" => Ok(CxxPolkadotChainMetadata {
+            balances_pallet_index: POLKADOT_TESTNET_BALANCES_PALLET,
+            transfer_allow_death_call_index: POLKADOT_TESTNET_TRANSFER_ALLOW_DEATH_CALL_INDEX,
+        }),
+        "Westend Asset Hub" => Ok(CxxPolkadotChainMetadata {
+            balances_pallet_index: POLKADOT_ASSET_HUB_TESTNET_BALANCES_PALLET,
+            transfer_allow_death_call_index:
+                POLKADOT_ASSET_HUB_TESTNET_TRANSFER_ALLOW_DEATH_CALL_INDEX,
+        }),
+        "Polkadot Asset Hub" => Ok(CxxPolkadotChainMetadata {
+            balances_pallet_index: POLKADOT_ASSET_HUB_MAINNET_BALANCES_PALLET,
+            transfer_allow_death_call_index:
+                POLKADOT_ASSET_HUB_MAINNET_TRANSFER_ALLOW_DEATH_CALL_INDEX,
+        }),
+        "Polkadot" => Ok(CxxPolkadotChainMetadata {
+            balances_pallet_index: POLKADOT_MAINNET_BALANCES_PALLET,
+            transfer_allow_death_call_index: POLKADOT_MAINNET_TRANSFER_ALLOW_DEATH_CALL_INDEX,
+        }),
+        _ => Err(Error::ChainNameNotFound),
+    };
+
+    Box::new(CxxPolkadotChainMetadataResult(metadata))
+}
+
+fn encode_unsigned_transfer_allow_death(
+    chain_metadata: &CxxPolkadotChainMetadata,
+    send_amount_bytes: &[u8; 16],
+    pubkey: &[u8; 32],
+) -> Vec<u8> {
+    let CxxPolkadotChainMetadata { balances_pallet_index, transfer_allow_death_call_index } =
+        chain_metadata;
+
+    let mut buf = Vec::<u8>::with_capacity(256);
```

Comment:
it strikes me that everything here is fixed size and could be done without allocating, although perhaps that's a premature optimization :sweat_smile:

Reply:
Maybe the 256 is misleading.

The encoded extrinsics are actually variable length, which makes using fixed-sized types painful and we can't know the size ahead of time.

I just chose 256 as a sane default to avoid frequent reallocations when appending to the Vec (as almost all instances of this extrinsic will fit in that). I think in Rust, it chooses a somewhat small default capacity based and then doubles capacity every time after. I just wanted to avoid that step.